e2e/tests/core/client_test.go

```diff
@@ -119,3 +138,206 @@ func (s *ClientTestSuite) TestClientUpdateProposal_Succeeds() {
+		isFullNode := strings.Contains(container.Names[0], fmt.Sprintf("%s-fn", chain.Config().ChainID))
+		if isNodeForDifferentChain || isFullNode {
+			continue
+		}
+
+		validatorPrivKey := fmt.Sprintf("/var/cosmos-chain/%s/config/priv_validator_key.json", chain.Config().Name)
+		privKeyFileContents, err := dockerutil.GetFileContentsFromContainer(ctx, s.DockerClient, container.ID, validatorPrivKey)
+		s.Require().NoError(err)
+
+		var filePV privval.FilePVKey
+		err = tmjson.Unmarshal(privKeyFileContents, &filePV)
+		s.Require().NoError(err)
+		filePvs = append(filePvs, filePV)
+	}
+
+	// We sort by address as GetValidatorSetByHeight also sorts by address. When iterating over them, the index
+	// will correspond to the correct ibcmock.PV.
+	sort.SliceStable(filePvs, func(i, j int) bool {
+		return filePvs[i].Address.String() < filePvs[j].Address.String()
+	})
+
+	for _, filePV := range filePvs {
+		pvs = append(pvs, &ibcmock.PV{
+			PrivKey: &ed25519.PrivKey{Key: filePV.PrivKey.Bytes()},
+		})
+	}
+
+	return pvs
+}
+
+// createMaliciousTMHeader creates a header with the provided trusted height with an invalid app hash.
+func createMaliciousTMHeader(
+	chainID string,
+	blockHeight int64,
+	trustedHeight clienttypes.Height,
+	timestamp time.Time,
+	tmValSet, tmTrustedVals *tmtypes.ValidatorSet,
+	signers []tmtypes.PrivValidator,
+	oldHeader *tmproto.Header,
+) (*ibctm.Header, error) {
+	tmHeader := tmtypes.Header{
+		Version:            tmprotoversion.Consensus{Block: tmversion.BlockProtocol, App: 2},
+		ChainID:            chainID,
+		Height:             blockHeight,
+		Time:               timestamp,
+		LastBlockID:        ibctesting.MakeBlockID(make([]byte, tmhash.Size), 10_000, make([]byte, tmhash.Size)),
+		LastCommitHash:     oldHeader.LastCommitHash,
+		ValidatorsHash:     tmValSet.Hash(),
+		NextValidatorsHash: tmValSet.Hash(),
+		DataHash:           tmhash.Sum([]byte(invalidHashValue)),
+		ConsensusHash:      tmhash.Sum([]byte(invalidHashValue)),
+		AppHash:            tmhash.Sum([]byte(invalidHashValue)),
+		LastResultsHash:    tmhash.Sum([]byte(invalidHashValue)),
+		EvidenceHash:       tmhash.Sum([]byte(invalidHashValue)),
+		ProposerAddress:    tmValSet.Proposer.Address, //nolint:staticcheck
+	}
+
+	hhash := tmHeader.Hash()
+	blockID := ibctesting.MakeBlockID(hhash, 3, tmhash.Sum([]byte(invalidHashValue)))
+	voteSet := tmtypes.NewVoteSet(chainID, blockHeight, 1, tmproto.PrecommitType, tmValSet)
+
+	commit, err := tmtypes.MakeCommit(blockID, blockHeight, 1, voteSet, signers, timestamp)
+	if err != nil {
+		return nil, err
+	}
+
+	signedHeader := &tmproto.SignedHeader{
+		Header: tmHeader.ToProto(),
+		Commit: commit.ToProto(),
+	}
+
+	valSet, err := tmValSet.ToProto()
+	if err != nil {
+		return nil, err
+	}
+
+	trustedVals, err := tmTrustedVals.ToProto()
+	if err != nil {
+		return nil, err
+	}
+
+	// The trusted fields may be nil. They may be filled before relaying messages to a client.
```

Comment:
sorry, this can be deleted as well. We are filling in the trusted fields as there is no relayer here
```suggestion
```